File: baekjoon/data_structure/5397/stack.py
import sys
from collections import deque


class Logger:
    def __init__(self, input_):
        self.left = deque()
        self.right = deque()

        speciallist = '<>-'
        funclist = [Logger.move_left, Logger.move_right, Logger.remove]
        for ch in input_:
            i = speciallist.find(ch)
            if i == -1:
                self.left.append(ch)
            else:
                funclist[i](self)

    def move_left(self):
        if self.left:
            self.right.appendleft(self.left[-1])
            self.left.pop()

    def move_right(self):
        if self.right:
            self.left.append(self.right[0])
            self.right.popleft()

    def remove(self):
        if self.left:
            self.left.pop()

    def output(self):
        return ''.join(self.left) + ''.join(self.right)


if __name__ == '__main__':
    def readline(): return sys.stdin.readline()

    test_cnt = int(readline())
    while test_cnt:
        vscode = Logger(readline().strip())
        print(vscode.output())
        test_cnt -= 1

File: baekjoon/data_structure/5397/stack.cpp
#include <iostream>
#include <deque>

typedef std::deque<int> t_deque;

class Logger
{
    private:
        t_deque left, right;

    public:
        void    take_input(char ch)
        {
            switch (ch)
            {
            case '<':
                if (!left.empty())
                {
                    right.push_front(left.back());
                    left.pop_back();
                }
                break ;
            case '>':
                if (!right.empty())
                {
                    left.push_back(right.front());
                    right.pop_front();
                }
                break ;
            case '-':
                if (!left.empty())
                    left.pop_back();
                break ;
            default:
                left.push_back(ch);
            }
        }

        std::string output(void)
        {
            return std::string(left.begin(), left.end())
                    + std::string(right.begin(), right.end());
        }
};

int main(void)
{
    std::cin.tie(0)->sync_with_stdio(0);

    int testCnt;
    std::string input;

    std::cin >> testCnt;
    while (testCnt)
    {
        std::cin >> input;

        Logger vscode;
        for (const char &ch : input)
            vscode.take_input(ch);

        std::cout << vscode.output() << '\n';
        --testCnt;
    }
    return (0);
}
